### ISSUE TITLE
[symm_mem] one/two-shot all-reduce deadlocks when issued from multiple streams: ops lack channel plumbing

### ISSUE THREAD

**Issue by lucifer1004** [pytorch/pytorch#189228]
### 🐛 Describe the bug

`symm_mem::two_shot_all_reduce_` (and the one-shot / multimem variants) synchronize the group with barriers that require the **device execution order of the ops to match their group issue order**. CUDA does not order kernels across streams, so issuing these ops from more than one stream — combined with realistic host-side skew between ranks — deadlocks the whole group permanently: all GPUs spin at 100%, no error is raised, launch queues fill up behind the spinning kernel.

The symmetric memory docs design **synchronization channels** for exactly this (barrier/put_signal/wait_signal take a `channel` argument with non-overlapping signal-pad slots, "so that a single handle can be used safely on multiple concurrent streams"): https://docs.pytorch.org/docs/main/symmetric_memory.html

However the high-level all-reduce ops neither expose a channel argument nor document a single-stream requirement:

```
symm_mem::two_shot_all_reduce_(Tensor(a!) input, str reduce_op, str group_name) -> Tensor(a!)
symm_mem::one_shot_all_reduce(Tensor input, str reduce_op, str group_name) -> Tensor
symm_mem::multimem_all_reduce_(Tensor(a!) input, str reduce_op, str group_name) -> Tensor(a!)
```

(their docstrings are empty; `CUDASymmetricMemoryOps.cu` on current main contains no channel plumbing).

**Minimal repro** (4 GPUs; `unordered` deadlocks in < 50 iterations, `ordered` — identical except for event fencing — runs 300 iterations bitwise-correct):

```python
# torchrun --nproc-per-node 4 repro.py [unordered|ordered]
import os, sys, time
import torch
import torch.distributed as dist
import torch.distributed._symmetric_memory as symm_mem

def main():
    rank = int(os.environ["RANK"])
    world = int(os.environ["WORLD_SIZE"])
    dist.init_process_group(backend="nccl")
    torch.cuda.set_device(rank)
    device = torch.device(f"cuda:{rank}")
    mode = sys.argv[1] if len(sys.argv) > 1 else "unordered"

    buf = symm_mem.empty(32 * 4096, dtype=torch.bfloat16, device=device)
    symm_mem.rendezvous(buf, dist.group.WORLD.group_name)

    torch.manual_seed(1234)
    x = torch.randn(32 * 4096, dtype=torch.bfloat16, device=device)
    golden = x.clone()
    dist.all_reduce(golden)
    torch.cuda.synchronize()

    s1, s2 = torch.cuda.Stream(), torch.cuda.Stream()
    for i in range(300):
        # stagger ranks, mimicking host-side pauses (compilation, tuning)
        if rank == i % world and i % 7 == 0:
            time.sleep(0.02)
        st = s1 if i % 2 == 0 else s2
        if mode == "ordered":
            st.wait_stream(torch.cuda.current_stream())
        with torch.cuda.stream(st):
            buf.copy_(x)
            torch.ops.symm_mem.two_shot_all_reduce_(
                buf, "sum", dist.group.WORLD.group_name
            )
        if mode == "ordered":
            torch.cuda.current_stream().wait_stream(st)
        if i % 50 == 0:
            torch.cuda.synchronize()
            diff = (buf.float() - golden.float()).abs().max().item()
            if rank == 0:
                print(f"iter {i}: max_diff={diff:.3e}", flush=True)
    torch.cuda.synchronize()
    dist.destroy_process_group()
    if rank == 0:
        print("DONE")

if __name__ == "__main__":
    main()
```

**How we hit it**: vLLM serving (TP=4). The engine's startup phases (memory profiling, torch.compile warmup, per-range warmup) issue collectives from different streams; an Nsight Systems capture shows the same process group's collectives on three different streams with second-scale rank skew from compilation pauses, several 1–1.5 s near-miss barrier spins, and then a permanent hang (GPUs pinned at 100%, `sample_tokens` RPC timeout, workers alive but unresponsive). Notably NCCL survives the identical multi-stream pattern in the same trace — its LL protocol matches by per-op flag slots rather than a group phase counter, which suggests the fix direction.

**Expected behavior** — any of:
1. The all-reduce ops expose a `channel` argument (matching the documented channel design), or
2. the single-stream-per-group contract is documented on the ops and enforced with a debug assertion, or
3. a channel is derived internally per stream.

**Workaround (validated)**: serialize all of a group's symm-mem ops on one dedicated stream at the caller, fenced with `wait_stream` events against the issuing stream — the `unordered` repro becomes bitwise-clean with this applied inside the caller.

### Versions

- torch 2.11.0+cu130 (also verified unaffected/unfixed on current main: `torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemoryOps.cu` has no channel plumbing)
- CUDA 13, driver 595.58.03, NCCL 2.28.9
- 4x NVIDIA RTX PRO 6000 Blackwell (SM120), PCIe P2P (no NVLink), Linux x86_64

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy

**Comment by kwen2501** [comment]
Today it is recommended that the user serialize the ops if they need to touch the same symm tensor and thus the same signal pad.
For `channel`, I am not sure if it could be used for stream -- today each op may use certain number of CTAs, each CTA would need signal pad space. So `channel=2` standing for `stream=2` may contend with an op that itself uses more than two CTAs.
cc @RohitRathore1

**Comment by RohitRathore1** [comment]
i think then we should add warning message in `CUDASymmetricMemoryOps` that will track the last stream used per group and will fire `TORCH_WARN_ONCE` if a different stream is detected and wire it into `two_shot_all_reduce_`, `one_shot_all_reduce`, `multimem_all_reduce_`, `multimem_all_gather_out`, and `reduce_scatter_out` and also documented the single-stream requirement with a `.. warning::` block in the docstrings for each of those ops in our docs, including the workaround (serialize onto a dedicated stream with `wait_stream` fencing). This way users hitting the vLLM pattern will get an actionable warning before they see a hang, and the contract is visible in the docs without any API change. I will be sending my PR soon.
cc: @kwen2501
